source/slang/slang-check-conversion.cpp

```diff
@@ -35,81 +35,101 @@ namespace Slang
-        if(as<MatrixExpressionType>(type)) return false;
+        if (as<ArrayExpressionType>(type)) return false;
+        if (as<VectorExpressionType>(type)) return false;
+        if (as<MatrixExpressionType>(type)) return false;
 
-        if(as<BasicExpressionType>(type))
+        if (as<BasicExpressionType>(type))
         {
             return true;
         }
 
-        if(as<ResourceType>(type))
+        if (as<ResourceType>(type))
         {
             return true;
         }
-        if(as<UntypedBufferResourceType>(type))
+        if (as<UntypedBufferResourceType>(type))
         {
             return true;
         }
-        if(as<SamplerStateType>(type))
+        if (as<SamplerStateType>(type))
         {
             return true;
         }
 
-        if(auto declRefType = as<DeclRefType>(type))
+        if (auto declRefType = as<DeclRefType>(type))
         {
-            if(as<StructDecl>(declRefType->getDeclRef()))
+            if (as<StructDecl>(declRefType->getDeclRef()))
                 return false;
         }
 
         return true;
     }
 
+    bool isSameSimpleExprType(Type* type1, Type* type2)
+    {
+        if (as<BasicExpressionType>(type1) && as<BasicExpressionType>(type2)) return true;
+        if (as<ArrayExpressionType>(type1) && as<ArrayExpressionType>(type2)) return true;
+        if (as<VectorExpressionType>(type1) && as<VectorExpressionType>(type2)) return true;
+        if (as<MatrixExpressionType>(type1) && as<MatrixExpressionType>(type2)) return true;
+        return false;
+    }
+
     bool SemanticsVisitor::shouldUseInitializerDirectly(
-        Type*    toType,
-        Expr*    fromExpr)
+        Type* toType,
+        Expr* fromExpr)
     {
         // A nested initializer list should always be used directly.
         //
-        if(as<InitializerListExpr>(fromExpr))
+        if (as<InitializerListExpr>(fromExpr))
         {
             return true;
         }
 
         // If the desired type is a scalar, then we should always initialize
         // directly, since it isn't an aggregate.
         //
-        if(isEffectivelyScalarForInitializerLists(toType))
+        if (isEffectivelyScalarForInitializerLists(toType))
             return true;
 
         // If the type we are initializing isn't effectively scalar,
         // but the initialization expression *is*, then it doesn't
         // seem like direct initialization is intended.
         //
-        if(isEffectivelyScalarForInitializerLists(fromExpr->type))
+        if (isEffectivelyScalarForInitializerLists(fromExpr->type))
             return false;
 
-        // Once the above cases are handled, the main thing
-        // we want to check for is whether a direct initialization
-        // is possible (a type conversion exists).
-        //
-        return canCoerce(toType, fromExpr->type, fromExpr);
+        // If 2 types are "broadly equal" we know the types can be coerced directly
```

Comment:
I want to understand why this change here, along with the change in line 151 is needed.

I think it should be fine if we revert all changes to this function and in 151.

If things break, I want to understand why.